### DISCUSSION TITLE
Question: Is it possible define minimum and maximum limits for date/datetime

### DISCUSSION THREAD

**Discussion by Havunen** [OAI/OpenAPI-Specification#2827] [Q&A] [answered]
Hi,

I have been trying to figure out how I could add limits to datetime / date based properties / values in open api 3 schema.

https://swagger.io/docs/specification/data-models/data-types/

describes that string can have min length and max length properties but I would like to provide minimum and maximum value for the date

```
type: string
minLength: 3
maxLength: 20
```

Something similar what integer has:

```
type: integer
minimum: 1
maximum: 20
```

Something like this?

```
type: string
format: date
minimum: 1996-12-19
maximum: 2021-12-19
```

**Comment by webron** [comment]
The short answer is 'no', as validation modifiers can be at most restricted to specific types, but not formats. That said, you can still add a regex to validate a given date range.

**Reply by rosdi** [reply]
A regex might work if min/max date is constant. It won't work if the date is dynamic. It is common for application to require date to be in the future, or date cannot be less than and/or exceed certain days/week/month/year from current

**Comment by flaksp** [comment, marked as answer]
As webron said it's not possible to add such constraints using OpenAPI or JSON Schema as-is. Only with hacks like using `pattern` keyword, or you should redesign your API to use UNIX timestamps so you will be able to use `minimum` and `maximum` keywords.

However JSON Schema specification says it's pretty valid to add custom keywords like this:

```yaml
type: string
format: date
minimumDate: 1996-12-19
maximumDate: 2021-12-19
```

Custom keywords `minimumDate` and `maximumDate` will be ignored (treated as annotations) by most tools. But if tools you are using support some kind of extensions or plugins you may "tune" them to recognise these keywords and you may write your own logic around this. If your tools don't have such API for extensions you may just fork them if these keywords worth this 😁

**Reply by handrews** [reply]
If you're using OAS 3.1 and a JSON Schema implementation that supports the standard output formats, you can read those `minimumDate` and `maximumDate` keywords in the output, along with the location of the data in the instance, and perform additional validation that way.

**Comment by handrews** [comment]
This seems to have been answered pretty thoroughly, closing.
